packages/grafana-data/src/transformations/transformers/insertValues.ts

```diff
@@ -0,0 +1,141 @@
+import { DataFrame, FieldType, Field, ArrayVector } from '../..';
+
+import { map } from 'rxjs/operators';
+
+import { DataTransformerID } from './ids';
+import { DataTransformerInfo } from '../../types/transformations';
+
+const enum InsertValuesMode {
+  Interval = 'interval',
+  Threshold = 'threshold',
+}
+
+const IS_TIME_FIELD = (field: Field) => field.type === FieldType.time;
+const THRESHOLD_HOUR = 3600 * 1e3;
+
+export interface InsertValuesTransformerOptions {
+  mode: InsertValuesMode;
+  // value to insert
+  value: any;
+  // finds field to use for interval or threshold checking
+  isRefField?: (field: Field) => boolean;
+  // minimum threshold required between adjacent non-null/non-undefined values
+  threshold?: number;
+  // should refField's values be scanned to find minimum interval?
+  guessInterval?: boolean;
+}
+
+export const insertValuesTransformer: DataTransformerInfo<InsertValuesTransformerOptions> = {
+  id: DataTransformerID.insertValues,
+  name: 'Insert Values',
+  description: 'Can be used to insert null values between adjacent points spanning a minimum time threshold',
+  defaultOptions: {
+    mode: InsertValuesMode.Interval,
+    value: null,
+    isRefField: IS_TIME_FIELD,
+    guessInterval: false,
+  },
+  operator: ({ mode, value = null, isRefField = IS_TIME_FIELD, guessInterval = false, threshold = THRESHOLD_HOUR }) => (
+    source
+  ) =>
+    source.pipe(
+      map((dataFrames) => {
+        if (mode === InsertValuesMode.Interval) {
+          return dataFrames.map((f) => insertValuesInterval(f, value ?? null, guessInterval ?? false, isRefField));
+        } else {
+          return dataFrames.map((f) => insertValuesThreshold(f, value ?? null, threshold, isRefField ?? IS_TIME_FIELD));
+        }
+      })
+    ),
+};
+
+// expects ascending number values in refField
+function getInterval(refField: Field, guessFromValues = false) {
+  let interval = Math.max(refField.config?.interval ?? 0, 0);
+
+  if (interval === 0 && guessFromValues && refField.values.length > 2) {
+    let refVals = refField.values.toArray();
+
+    interval = Infinity;
+
+    for (let i = 1; i < refVals.length; i++) {
+      interval = Math.min(interval, refVals[i] - refVals[i - 1]);
+    }
+  }
+
+  return Number.isFinite(interval) ? interval : 0;
+}
+
+// assumes some interval steps may be absent, but what remains in the time field:
+// 1. contains only numbers, would pass Number.isFinite(value)
+// 2. conforms to the specified interval
```

Comment:
i would love to be able to not have assumption [2].

based on what i know, the current version of Prometheus and Loki does fulfil [2], but it feels to me we are relying a little too much on some implementation detail.

also, this might be useful later for other datasources too, ones where there is no such guarantee.

i think an algorithm which finds-gaps and fills-gaps would be a lot more robust.

i understand that the approach described here allows for faster code, but i'm worried that later i'll be debugging tricky user-data coming from some-kind-of-proxy or some-kind-of-basically-mostly-prometheus-but-not-completely. i'll help in any way needed to find a fast-enough find-gaps-fill-gaps algo, if that's acceptable. (maybe something that does a first-pass check if there are gaps, and if no gaps just returns the orig data (i guess most data has no gaps) etc.)

an example of what-i-am-worried-about: when sending a query to Prometheus, i can send the `interval` as a float-number or a hmm.. "prometheus string", like "500ms". if i send it as a float, i ASSUME Prometheus rounds it to milliseconds and then works with it as millisecond-as-int. so if you send it in as a float, you have to hope Go rounds the same way as Javascript.  of course, in grafana, we can fix it by just never sending interval-as-float, but i hope this shows the problem.

Reply:
(if we must keep assumption [2] (i hope not 😄 ), then maybe when the algorithm starts, run through the time-field and verify the assumption? should be fast-enough.. question is, what to do when we find out that the time-field is not "good"... i'd vote for return-error)

Reply:
i've re-implemented this to remove assumption [2]. it will now insert single `null` values at midpoint timestamps between any adjacent timestamps > threshold, where threshold is pulled from time field's `config.interval`, when present and not explicitly defined in the transformer opts.